v2/yarpcrouter/router.go

```diff
@@ -58,19 +58,66 @@ type MapRouter struct {
 func NewMapRouter(defaultService string) MapRouter {
 	return MapRouter{
 		defaultService:            defaultService,
-		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.TransportProcedure),
 		serviceNames:              map[string]struct{}{defaultService: {}},
 	}
 }
 
+// NewMapRouterWithEncodingProcedures constructs a new MapRouter with the given default service name and registers
+// the given encoding level procedures as transport level procedures
+func NewMapRouterWithEncodingProcedures(defaultService string, encodingProcedures []yarpc.EncodingProcedure) MapRouter {
+	router := MapRouter{
+		defaultService:            defaultService,
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]yarpc.TransportProcedure),
+		serviceNames:              map[string]struct{}{defaultService: {}},
+	}
+
+	transportProcedures := []yarpc.TransportProcedure{}
+	for _, p := range encodingProcedures {
+		handler := func(c context.Context, r *yarpc.Request, b *yarpc.Buffer) (*yarpc.Response, *yarpc.Buffer, error) {
```

Comment:
This’ll work!

We use HandlerFunc in tests a lot. We prefer objects with distinguished methods over closures for production, mostly on account of being more inspectable, e.g., if we surfaced a debug page of all registered procedures.